### ISSUE TITLE
VerifyError: Bad type on operand stack in arraylength

### ISSUE THREAD

**Issue by christianbalzer** [powermock/powermock#784]
All of a sudden, a PowerMock test started failing today, with the following exception:

> `java.lang.VerifyError: Bad type on operand stack in arraylength`

I'm using Java 1.8.0.121 on a Mac, Mockito 1.10.19 and PowerMock 1.6.4.
We are also using [springockito](https://github.com/springockito/springockito), so can't upgrade to Mockito 2.

Here is a more complete stack trace:

```
java.lang.VerifyError: Bad type on operand stack in arraylength
Exception Details:
  Location:
    com.example.FooTest.bar()V @852: arraylength
  Reason:
    Invalid type: 'java/lang/Object' (current frame, stack[0])
  Current Frame:
    bci: @852
    flags: { }
    locals: { 'com.example.FooTest', 'java/lang/String', top, 'java/lang/Class', '[Ljava/lang/String;', integer, integer, 'java/lang/String', 'java/lang/Object', top, top, top, top, top, 'java/lang/Object' }
    stack: { 'java/lang/Object' }
  Bytecode:
    [...]
  Stackmap Table:
    [...]
  at java.lang.Class.getDeclaredMethods0(Native Method)
	at java.lang.Class.privateGetDeclaredMethods(Class.java:2701)
	at java.lang.Class.privateGetPublicMethods(Class.java:2902)
	at java.lang.Class.getMethods(Class.java:1615)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.getTestMethods(PowerMockJUnit44RunnerDelegateImpl.java:93)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit44RunnerDelegateImpl.<init>(PowerMockJUnit44RunnerDelegateImpl.java:69)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit47RunnerDelegateImpl.<init>(PowerMockJUnit47RunnerDelegateImpl.java:42)
	at org.powermock.modules.junit4.internal.impl.PowerMockJUnit49RunnerDelegateImpl.<init>(PowerMockJUnit49RunnerDelegateImpl.java:25)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.createDelegatorFromClassloader(JUnit4TestSuiteChunkerImpl.java:156)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.createDelegatorFromClassloader(JUnit4TestSuiteChunkerImpl.java:40)
	at org.powermock.tests.utils.impl.AbstractTestSuiteChunkerImpl.createTestDelegators(AbstractTestSuiteChunkerImpl.java:244)
	at org.powermock.modules.junit4.common.internal.impl.JUnit4TestSuiteChunkerImpl.<init>(JUnit4TestSuiteChunkerImpl.java:61)
	at org.powermock.modules.junit4.common.internal.impl.AbstractCommonPowerMockRunner.<init>(AbstractCommonPowerMockRunner.java:32)
	at org.powermock.modules.junit4.PowerMockRunner.<init>(PowerMockRunner.java:34)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance0(Native Method)
	at sun.reflect.NativeConstructorAccessorImpl.newInstance(NativeConstructorAccessorImpl.java:62)
	at sun.reflect.DelegatingConstructorAccessorImpl.newInstance(DelegatingConstructorAccessorImpl.java:45)
	at java.lang.reflect.Constructor.newInstance(Constructor.java:423)
	at org.junit.internal.builders.AnnotatedBuilder.buildRunner(AnnotatedBuilder.java:29)
	at org.junit.internal.builders.AnnotatedBuilder.runnerForClass(AnnotatedBuilder.java:21)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.builders.AllDefaultPossibilitiesBuilder.runnerForClass(AllDefaultPossibilitiesBuilder.java:26)
	at org.junit.runners.model.RunnerBuilder.safeRunnerForClass(RunnerBuilder.java:59)
	at org.junit.internal.requests.ClassRequest.getRunner(ClassRequest.java:26)
	at org.junit.internal.requests.FilterRequest.getRunner(FilterRequest.java:31)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:49)
	at com.intellij.rt.execution.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:51)
	at com.intellij.rt.execution.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:242)
	at com.intellij.rt.execution.junit.JUnitStarter.main(JUnitStarter.java:70)
```

The test class, `FooTest`, is annotated with `@RunWith(PowerMockRunner.class)` and the test with `@Test` only. Anther test in the class is also annotated with `@PrepareOnlyThisForTest`

Here is the test:

```
@Test
    public void bar() {
        final String propertyAccessorBaseName = myPropertyName.substring(0, 1).toUpperCase() + myPropertyName.substring(1);
        final String[] SETTER_PREFIXES = { "set", "with" };

        Class myClass = Baz.class;
        for(String prefix : SETTER_PREFIXES) {
            for (Method method : myClass.getDeclaredMethods()) {
                String getterName = prefix + propertyAccessorBaseName;
                if (getterName.equals(method.getName())) {
                    fail(myPropertyName + " should be read only for the front-end, but found setter: " + getterName);
                }
            }
        }
    }
```

This SO post _might_ be related: http://stackoverflow.com/a/25395106/2018047

**Comment by thekingnothing** [comment]
> All of a sudden, a PowerMock test started failing today, with the following exception:
What have happened that test started failing? Have something changed? Java version? new library? Which version of Javassist is used? 

PowerMock uses Javassist for byte code modification. So version of Javassist is important.

**Comment by christianbalzer** [comment]
`mvn dependency:tree` produced the following:

```
[INFO] +- javassist:javassist:jar:3.12.1.GA:compile
...
[INFO] +- org.powermock:powermock-module-junit4:jar:1.6.4:test
[INFO] |  \- org.powermock:powermock-module-junit4-common:jar:1.6.4:test
[INFO] |     +- org.powermock:powermock-core:jar:1.6.4:test
[INFO] |     |  \- org.javassist:javassist:jar:3.20.0-GA:test
[INFO] |     \- org.powermock:powermock-reflect:jar:1.6.4:test
```
So, both `3.12.1.GA` and `3.20.0-GA` are on the classpath...

**Comment by thekingnothing** [comment]
Could you try to use the latest version of Javassist? Upgrade version in your  pom and add PowerMock dependencies to exclude.

On 2 May 2017, at 17:45, christianbalzer <notifications@github.com<mailto:notifications@github.com>> wrote:


mvn dependency:tree produced the following:

[INFO] +- javassist:javassist:jar:3.12.1.GA:compile
...
[INFO] +- org.powermock:powermock-module-junit4:jar:1.6.4:test
[INFO] |  \- org.powermock:powermock-module-junit4-common:jar:1.6.4:test
[INFO] |     +- org.powermock:powermock-core:jar:1.6.4:test
[INFO] |     |  \- org.javassist:javassist:jar:3.20.0-GA:test
[INFO] |     \- org.powermock:powermock-reflect:jar:1.6.4:test


So, both 3.12.1.GA and 3.20.0-GA are on the classpath...

**Comment by christianbalzer** [comment]
Sorry for the delay. The immediate problem was fixed by updating to `javassist` `3.21.0-GA`.

However, I now get 
```
java.lang.Exception: No tests found matching Method bar(com.example.FooTest) from 
org.junit.internal.requests.ClassRequest@10e31a9a
	at org.junit.internal.requests.FilterRequest.getRunner(FilterRequest.java:35)
```

The test method is annotated with `@PrepareOnlyThisForTest(Static.class)`

If I move that annotation to the class level, it works, though.

Shall I raise another ticket about that "filter" issue?

**Comment by thekingnothing** [comment]
Yeah, please raise a new issue for this case.

I afraid that, only `@PreapareForTest` annotation is tested on method level.

On 9 May 2017, at 13:09, Christian Balzer <notifications@github.com<mailto:notifications@github.com>> wrote:


Sorry for the delay. The immediate problem was fixed by updating to javassist 3.21.0-GA.

However, I now get

java.lang.Exception: No tests found matching Method bar(com.example.FooTest) from
org.junit.internal.requests.ClassRequest@10e31a9a
        at org.junit.internal.requests.FilterRequest.getRunner(FilterRequest.java:35)


The test method is annotated with @PrepareOnlyThisForTest(Static.class)

If I move that annotation to the class level, it works, though.

Shall I raise another ticket about that "filter" issue?

**Comment by windqdu** [comment]
**我也遇到该问题了：**

java.lang.VerifyError: Bad type on operand stack in arraylength
Exception Details:
  Location:
    com/hisense/hitv/diffcreater/api/utils/CommonParamUtils.getClientIP(Ljavax/servlet/http/HttpServletRequest;)Ljava/lang/String; @1466: arraylength
  Reason:
    Invalid type: 'java/lang/Object' (current frame, stack[0])
  Current Frame:
    bci: @1466
    flags: { }
    locals: { 'javax/servlet/http/HttpServletRequest', 'java/lang/String', 'java/lang/Object', top, top, top, top, top, top, 'java/lang/Object' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: 1304 70b8 0472 1304 7304 bd00 3059 032a
    0000010: 5313 0475 b804 7713 0478 b804 7a4c 2b01
    0000020: 3a07 013a 0819 0713 0d8f b80d 9113 0d92
    0000030: 130d 93b8 0d95 b80d 973a 0919 09b2 0d99

后来经过一番折腾，终于发现问题所在。
我使用是的powermock-easymock-junit-1.7.1,但是javassist.jar用的是javassist-3.12.0.GA.jar，实际上powerMock1.7.1依赖的是javassist-3.21.0-GA.jar，删除了javassist-3.12.0.GA.jar，并添加了javassist-3.21.0-GA.jar，该错误就消失了。

该问题也可能是由其他依赖的jar包冲突导致，如果javassist.jar版本没有问题，可仔细检查其他依赖的jar
